src/ray/gcs/gcs_server/gcs_task_manager.h

```diff
@@ -203,18 +245,152 @@ class GcsTaskManager : public rpc::TaskInfoHandler {
     void MarkTasksFailedOnWorkerDead(const WorkerID &worker_id,
                                      const rpc::WorkerTableData &worker_failure_data);
 
-   private:
-    /// Get a reference to the TaskEvent stored in the buffer.
+    /// Get the job task summary given a job id.
     ///
-    /// \param task_attempt The task attempt.
-    /// \return Reference to the task events stored in the buffer.
-    rpc::TaskEvents &GetTaskEvent(const TaskAttempt &task_attempt);
+    /// Caller should make sure the job id exists by calling HasJob() first.
+    ///
+    /// \param job_id Job ID.
+    const JobTaskSummary &GetJobTaskSummary(const JobID &job_id) const {
+      auto it = job_task_summary_.find(job_id);
+      RAY_CHECK(it != job_task_summary_.end());
+      return it->second;
+    }
+
+    void UpdateJobSummaryOnJobDone(const JobID &job_id) {
+      auto it = job_task_summary_.find(job_id);
+      RAY_CHECK(it != job_task_summary_.end());
+      it->second.OnJobEnds();
+    }
+
+    void GcJobSummary() {
```

Comment:
This is the major GC change

Reply:
hmm can you give me a summary what was the root cause & what is the fix?

Reply:
Root cause is:
- We are tracking the `dropped_task_attempts_` as we drop task attempts, but this set could go unbounded before the job finishes (since we were only GCing this set once the job is finished) 
- So when a job is long running + generating many task attempts, this will cause OOMs

Fix:
- Enforce a hard limit on how many dropped task attempts we track (default 1M ->  less than 100 MiB overhead) 
- As an optimization: we will evict those older task attempts dropped (assuming they are less likely to be updated)

TODOs:
- On top this periodic hard limit, we could also do better by stopping tracking a dropped task attempt when we know the task has entered a terminal state (e.g. FINISHED or FAILED). This will come as a follow-up.